File: backend/expiry_kind.py
"""Monthly vs weekly F&O expiry (NSE last Tuesday, BSE last Thursday)."""
from __future__ import annotations

import calendar
from datetime import date, datetime
from typing import Any, Iterable, List, Optional, Union

TUESDAY = 1
THURSDAY = 3

DateLike = Union[str, date, datetime]


def _as_date(raw: DateLike) -> Optional[date]:
    if isinstance(raw, datetime):
        return raw.date()
    if isinstance(raw, date):
        return raw
    s = str(raw or "")[:10]
    try:
        return date.fromisoformat(s)
    except ValueError:
        return None


def last_weekday_of_month(year: int, month: int, weekday: int) -> date:
    last = calendar.monthrange(year, month)[1]
    d = date(year, month, last)
    return date(year, month, last - ((d.weekday() - int(weekday)) % 7))


def monthly_expiry_weekday(index: Any) -> int:
    u = str(index or "").upper().replace(" ", "")
    if "SENSEX" in u or u == "BANKEX":
        return THURSDAY
    # NSE index options: last Tuesday. Other F&O (MCX, leftover BSE) keep Thursday.
    if "NIFTY" in u:
        return TUESDAY
    return THURSDAY


def is_monthly_expiry(raw: DateLike, index: Any = "NIFTY") -> bool:
    d = _as_date(raw)
    if d is None:
        return False
    wd = monthly_expiry_weekday(index)
    return d.weekday() == wd and d == last_weekday_of_month(d.year, d.month, wd)


def expiry_tag(raw: DateLike, index: Any = "NIFTY") -> str:
    return "M" if is_monthly_expiry(raw, index) else "W"


def annotate_expiries(dates: Iterable[DateLike], index: Any = "NIFTY", *, today: Optional[date] = None) -> List[dict]:
    today = today or date.today()
    out = []
    for raw in dates or []:
        d = _as_date(raw)
        if d is None:
            continue
        monthly = is_monthly_expiry(d, index)
        out.append({
            "date": d.isoformat(),
            "tag": "M" if monthly else "W",
            "type": "monthly" if monthly else "weekly",
            "days_to_expiry": (d - today).days,
            "label": d.strftime("%d %b").lstrip("0"),
        })
    return out

File: frontend/src/lib/expiryKind.js
/** Monthly vs weekly F&O expiry: NSE last Tuesday, BSE last Thursday. */

const TUESDAY = 2; // Date.getUTCDay(): Sun=0 … Tue=2
const THURSDAY = 4;

function ymdParts(iso) {
  const s = String(iso || "").slice(0, 10);
  const [y, m, d] = s.split("-").map(Number);
  if (!y || !m || !d) return null;
  return { y, m, d, iso: s };
}

export function monthlyExpiryWeekdayJs(index) {
  const u = String(index || "").toUpperCase().replace(/\s+/g, "");
  if (u.includes("SENSEX") || u === "BANKEX") return THURSDAY;
  if (u.includes("NIFTY")) return TUESDAY;
  return THURSDAY;
}

export function lastWeekdayOfMonthUtc(year, month, weekdaySun0) {
  const last = new Date(Date.UTC(year, month, 0)); // month is 1-12 → day 0 of next month
  const lastDay = last.getUTCDate();
  const lastWd = last.getUTCDay();
  const offset = (lastWd - weekdaySun0 + 7) % 7;
  const day = lastDay - offset;
  const dt = new Date(Date.UTC(year, month - 1, day));
  return dt.toISOString().slice(0, 10);
}

export function isMonthlyExpiry(iso, index = "NIFTY") {
  const p = ymdParts(iso);
  if (!p) return false;
  const want = monthlyExpiryWeekdayJs(index);
  const last = lastWeekdayOfMonthUtc(p.y, p.m, want);
  const dt = new Date(Date.UTC(p.y, p.m - 1, p.d));
  return dt.getUTCDay() === want && p.iso === last;
}

export function expiryTag(iso, index = "NIFTY") {
  return isMonthlyExpiry(iso, index) ? "M" : "W";
}

export function annotateExpiries(dates, index = "NIFTY", todayIso) {
  const today = ymdParts(todayIso) || ymdParts(new Date().toISOString());
  const t = today ? Date.UTC(today.y, today.m - 1, today.d) : Date.now();
  return (dates || [])
    .map((raw) => String(raw || "").slice(0, 10))
    .filter(Boolean)
    .map((iso) => {
      const p = ymdParts(iso);
      const monthly = isMonthlyExpiry(iso, index);
      let days = null;
      if (p) {
        days = Math.round((Date.UTC(p.y, p.m - 1, p.d) - t) / 86400000);
      }
      const label = p
        ? new Date(Date.UTC(p.y, p.m - 1, p.d)).toLocaleDateString("en-GB", {
            day: "numeric",
            month: "short",
            timeZone: "UTC",
          })
        : iso;
      return {
        date: iso,
        tag: monthly ? "M" : "W",
        type: monthly ? "monthly" : "weekly",
        days_to_expiry: days,
        label,
      };
    });
}
